overlord/snapstate/autorefresh.go

```diff
@@ -974,3 +974,133 @@ func MockRefreshCandidate(snapSetup *SnapSetup) interface{} {
 		SnapSetup: *snapSetup,
 	}
 }
+
+func incrementSnapRefreshFailures(st *state.State, snapsup *SnapSetup) error {
+	var snapst SnapState
+	err := Get(st, snapsup.InstanceName(), &snapst)
+	if err != nil {
+		return err
+	}
+
+	// Update refresh failure information for snap revision
+	if snapst.RefreshFailures != nil && snapst.RefreshFailures.ToRevision == snapsup.Revision() {
+		snapst.RefreshFailures.FailureCount++
+		snapst.RefreshFailures.LastFailureTime = timeNow()
+	} else {
+		snapst.RefreshFailures = &snap.RefreshFailuresInfo{
+			ToRevision:      snapsup.Revision(),
+			FailureCount:    1,
+			LastFailureTime: timeNow(),
+		}
+	}
+	Set(st, snapsup.InstanceName(), &snapst)
+
+	delay := computeSnapRefreshRemainingDelay(snapst.RefreshFailures).Round(time.Hour)
+	logger.Noticef("snap %q auto-refresh to revision %s has failed, next auto-refresh attempt will be delayed by %v hours", snapsup.InstanceName(), snapsup.Revision(), delay.Hours())
+	return nil
+}
+
+func processFailedAutoRefresh(chg *state.Change, _ state.Status, new state.Status) {
+	if chg.Kind() != "auto-refresh" || new != state.ErrorStatus {
+		return
+	}
+
+	var failedSnapNames []string
+	for _, t := range chg.Tasks() {
+		// We only care about snaps that failed after unlink-current-snap because
+		// this indicates (with high probability) that something related to the snap
+		// itself is broken.
+		if t.Kind() != "unlink-current-snap" || t.Status() != state.UndoneStatus {
+			continue
+		}
+
+		snapsup, err := TaskSnapSetup(t)
+		if err != nil {
+			logger.Debugf("internal error: failed to get snap associated with task %s: %v", t.ID(), err)
+			continue
+		}
+		if err := incrementSnapRefreshFailures(t.State(), snapsup); err != nil {
+			logger.Debugf("internal error: failed to increment failure count for snap %q: %v", snapsup.InstanceName(), err)
+			continue
+		}
+
+		failedSnapNames = append(failedSnapNames, snapsup.InstanceName())
+	}
+
+	if len(failedSnapNames) == 0 {
+		return
+	}
+
+	// Attach failed snaps to change api data. This intended to guide
+	// agents on devices that manage their own refresh cycle.
+	var data map[string]interface{}
+	err := chg.Get("api-data", &data)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		logger.Debugf("internal error: failed to get api-data for change %s: %v", chg.ID(), err)
+		return
+	}
+	if len(data) == 0 {
+		data = make(map[string]interface{})
+	}
+	data["refresh-failed"] = failedSnapNames
```

Comment:
do we sort snap-names? should we sort this?

Reply:
> do we sort snap-names?

For auto-refresh changes, yes. 

> should we sort this?

Sure, for consistency

Reply:
Updated

Reply:
isn't there sorting in the tests that now we can drop?

Reply:
done